Quarkus OIDC shouldn't fail, when Auth Server is not up

**Describe the bug**
When I start a quarkus application with the oidc extension used, the startup will fail, when the auth server is not reachable. 

**Expected behavior**
Application should start normal, maybe with a warning and an automatic retry

**Actual behavior**
Application stops with following exception:
`java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy_0(VertxKeycloakBuildStep$setup28.zig:92)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy(VertxKeycloakBuildStep$setup28.zig:36)
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:137)
	at io.quarkus.runtime.Application.start(Application.java:94)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:336)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:685)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
2019-10-16 20:28:40,692 ERROR [io.qua.dev.DevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:185)
	at io.quarkus.runtime.Application.start(Application.java:94)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy_0(VertxKeycloakBuildStep$setup28.zig:92)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy(VertxKeycloakBuildStep$setup28.zig:36)
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:137)
	... 5 more
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:336)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:685)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)`


**To Reproduce**
Steps to reproduce the behavior:
1. Build minimal application
2. Add oicd dependency
3. Add auth server address to application.properties
4. Start server

**Configuration**
```properties
# Add your application.properties here, if applicable.

`quarkus.oidc.auth-server-url=http://localhost:8180/auth/realms/test
quarkus.oidc.client-id=test-client`

Another error ocurres when the keycloak is up, however the REALM is incorrect. Is it a good idea open another Issue?

```
2019-10-16 19:06:46,165 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 1912ms
java.util.concurrent.CompletionException: io.vertx.core.impl.NoStackTraceThrowable: Not Found
        at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
        at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
        at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
        at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup44.deploy_0(VertxKeycloakBuildStep$setup44.zig:73)
        at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup44.deploy(VertxKeycloakBuildStep$setup44.zig:92)
        at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:169)
        at io.quarkus.runtime.Application.start(Application.java:94)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
        at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: io.vertx.core.impl.NoStackTraceThrowable: Not Found
2019-10-16 19:06:48,149 ERROR [io.qua.dev.DevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:225)
        at io.quarkus.runtime.Application.start(Application.java:94)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
        at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: java.util.concurrent.CompletionException: io.vertx.core.impl.NoStackTraceThrowable: Not Found
        at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
        at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
        at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
        at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup44.deploy_0(VertxKeycloakBuildStep$setup44.zig:73)
        at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup44.deploy(VertxKeycloakBuildStep$setup44.zig:92)
        at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:169)
        ... 5 more
Caused by: io.vertx.core.impl.NoStackTraceThrowable: Not Found

2019-10-16 19:06:48,150 ERROR [io.qua.dev.DevModeMain] (main) Failed to start Quarkus, attempting to start hot replacement endpoint to recover
```

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java
@@ -1,5 +1,6 @@
 package io.quarkus.oidc.runtime;
 
+import java.time.Duration;
 import java.util.List;
 import java.util.Optional;
 
@@ -49,6 +50,13 @@ public class OidcConfig {
      */
     @ConfigItem
     Optional<String> clientId;
+
+    /**
+     * The maximum amount of time the adapter will try connecting to the currently unavailable OIDC server for.
+     * For example, setting it to '20S' will let the adapter keep requesting the connection for up to 20 seconds.
+     */
+    @ConfigItem
+    public Optional<Duration> connectionDelay;
 
     /**
      * Configuration to find and parse a custom claim containing the roles information.

--- a/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java
+++ b/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java
@@ -2,6 +2,8 @@ package io.quarkus.oidc.runtime;
 
 import java.util.concurrent.CompletableFuture;
 
+import org.jboss.logging.Logger;
+
 import io.quarkus.arc.runtime.BeanContainer;
 import io.quarkus.oidc.OIDCException;
 import io.quarkus.runtime.RuntimeValue;
@@ -16,6 +18,8 @@ import io.vertx.ext.auth.oauth2.providers.KeycloakAuth;
 
 @Recorder
 public class OidcRecorder {
+
+    private static final Logger LOG = Logger.getLogger(OidcRecorder.class);
 
     public void setup(OidcConfig config, RuntimeValue<Vertx> vertx, BeanContainer beanContainer) {
         OAuth2ClientOptions options = new OAuth2ClientOptions();
@@ -45,19 +49,43 @@ public class OidcRecorder {
                     .setPublicKey(config.publicKey.get()));
         }
 
-        CompletableFuture<OAuth2Auth> cf = new CompletableFuture<>();
-        KeycloakAuth.discover(vertx.getValue(), options, new Handler<AsyncResult<OAuth2Auth>>() {
-            @Override
-            public void handle(AsyncResult<OAuth2Auth> event) {
-                if (event.failed()) {
-                    cf.completeExceptionally(toOidcException(event.cause()));
+        final long connectionDelayInSecs = config.connectionDelay.isPresent() ? config.connectionDelay.get().toMillis() / 1000
+                : 0;
+        final long connectionRetryCount = connectionDelayInSecs > 1 ? connectionDelayInSecs / 2 : 1;
+        if (connectionRetryCount > 1) {
+            LOG.infof("Connecting to IDP for up to %d times every 2 seconds", connectionRetryCount);
+        }
+
+        OAuth2Auth auth = null;
+        for (long i = 0; i < connectionRetryCount; i++) {
+            try {
+                CompletableFuture<OAuth2Auth> cf = new CompletableFuture<>();
+                KeycloakAuth.discover(vertx.getValue(), options, new Handler<AsyncResult<OAuth2Auth>>() {
+                    @Override
+                    public void handle(AsyncResult<OAuth2Auth> event) {
+                        if (event.failed()) {
+                            cf.completeExceptionally(toOidcException(event.cause()));
+                        } else {
+                            cf.complete(event.result());
+                        }
+                    }
+                });
+
+                auth = cf.join();
+                break;
+            } catch (OIDCException ex) {
+                if (i + 1 < connectionRetryCount) {
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException iex) {
+                        // continue connecting
+                    }
+
                 } else {
-                    cf.complete(event.result());
+                    throw ex;
                 }
             }
-        });
-
-        OAuth2Auth auth = cf.join();
+        }
 
         OidcIdentityProvider identityProvider = beanContainer.instance(OidcIdentityProvider.class);
         identityProvider.setAuth(auth);

